performance_manager/lib/l2_headways.py

```diff
@@ -0,0 +1,383 @@
+    )
+
+    # final select statement to be used for insertion of records into temporary
+    # headways table
+    headways_final_query = sa.select(
+        headways_lag_query.c.fk_vp_stopped_event,
+        headways_lag_query.c.fk_tu_stopped_event,
+        headways_lag_query.c.hash.label("fk_trip_event_hash"),
+        (
+            headways_lag_query.c.timestamp_start
+            - headways_lag_query.c.prev_timestamp_start
+        ).label("headway_seconds"),
+    )
+
+    insert_cols = [
+        "fk_vp_stopped_event",
+        "fk_tu_stopped_event",
+        "fk_trip_event_hash",
+        "headway_seconds",
+    ]
+
+    insert_query = (
+        TempHeadways.metadata.tables[TempHeadways.__tablename__]
+        .insert()
+        .from_select(
+            insert_cols,
+            headways_final_query,
+        )
+    )
+
+    db_manager.truncate_table(TempHeadways)
+    result = db_manager.execute(insert_query)
+
+    process_logger.add_metadata(temp_rows_inserted=result.rowcount)
+    process_logger.log_complete()
+
+    return result.rowcount
+
+
+def update_headways_from_temp(db_manager: DatabaseManager) -> None:
+    """
+    update headways tables with UPDATE from SELECT of TempHeadways table
+    """
+    process_logger = ProcessLogger("l2_headways_update")
+    process_logger.log_start()
+    update_query = (
+        Headways.metadata.tables[Headways.__tablename__]
+        .update()
+        .values(
+            fk_trip_event_hash=TempHeadways.fk_trip_event_hash,
+            fk_tu_stopped_event=TempHeadways.fk_tu_stopped_event,
+            fk_vp_stopped_event=TempHeadways.fk_vp_stopped_event,
+            headway_seconds=TempHeadways.headway_seconds,
+        )
+        .where(Headways.fk_trip_event_hash == TempHeadways.fk_trip_event_hash)
+    )
+
+    result = db_manager.execute(update_query)
+
+    process_logger.add_metadata(rows_updated=result.rowcount)
+    process_logger.log_complete()
+
+
+def insert_new_headways(db_manager: DatabaseManager) -> None:
+    """
+    insert new headways into Headways table with INSERT from SELECT of
+    TempHeadways table
+    """
+    insert_cols = [
+        "fk_vp_stopped_event",
+        "fk_tu_stopped_event",
+        "fk_trip_event_hash",
+        "headway_seconds",
+    ]
+
+    process_logger = ProcessLogger("l2_headways_insert")
+    process_logger.log_start()
+    insert_query = (
+        Headways.metadata.tables[Headways.__tablename__]
+        .insert()
+        .from_select(
+            insert_cols,
+            sa.select(
+                TempHeadways.fk_vp_stopped_event,
+                TempHeadways.fk_tu_stopped_event,
+                TempHeadways.fk_trip_event_hash,
+                TempHeadways.headway_seconds,
+            )
+            .join(
+                Headways,
+                Headways.fk_trip_event_hash == TempHeadways.fk_trip_event_hash,
+                isouter=True,
+            )
+            .where((Headways.fk_trip_event_hash.is_(None))),
```

Comment:
how does this part work? i would think that the fk_trip_event_hash in temp wouldn't be in headways yet.

Reply:
You are correct.

This query does a JOIN between `TempHeadways` and `Headways`, specifically a LEFT JOIN where `TempHeadways` is the left table, so all records from `TempHeadways` are included in the JOIN. Then in the WHERE clause we limit the results to records where `fk_trip_event_hash` from the `Headways` table is NULL:

`.where((Headways.fk_trip_event_hash.is_(None)))`

This gives us only the records from `TempHeadways` that don't exist in `Headways`